pytorch_lightning/loops/closure.py

```diff
@@ -0,0 +1,113 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Callable, Optional
+
+from torch import Tensor
+
+from pytorch_lightning.profiler import BaseProfiler
+from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+from pytorch_lightning.utilities.warnings import WarningCache
+
+
+@dataclass
+class ClosureResult:
+    closure_loss: Optional[Tensor]
+    result_collection: Optional[ResultCollection]
+
+    @property
+    def loss(self) -> Optional[Tensor]:
+        if self.closure_loss is not None:
+            return self.closure_loss.detach().clone()
+
+
+class Closure(ABC):
```

Comment:
Just for the naming: IMO this should be the `LightningClosure` (if having one at all), this would be consistent with `LightningEnums`, which is also the base there.

Reply:
For why I did two classes:

The idea here was that the Closure abstract class defines the minimal interface for it to mimic a traditional closure that we can pass to the optimizer_step but that implements the notion of "getting a result back out of it". 

What runs inside the closure is specific to lightning's implementation, therefore the separate class LightningClosure.

For the naming, DefaultLightningClosure, I'm not fully convinced since we don't have this naming convention in Lightning.

Reply:
i asked about this here: https://github.com/PyTorchLightning/pytorch-lightning/pull/8700/files#r687310014

we are not consistent today on components prefixed with Lightning vs not. i lean towards not having the prefix because it's already under the lightning path pytorch_lightning.loops.closure

Reply:
I renamed: Closure -> AbstractClosure, LightningClosure -> Closure